hefquin-engine/src/main/java/se/liu/ida/hefquin/engine/HeFQUINEngineImpl.java

```diff
@@ -30,14 +32,22 @@ public class HeFQUINEngineImpl implements HeFQUINEngine
 {
 	protected final FederationAccessManager fedAccessMgr;
 	protected final QueryProcessor qProc;
+	protected final ExecutorService execFed;
+	protected final ExecutorService execPlan;
 
 	public HeFQUINEngineImpl( final FederationAccessManager fedAccessMgr,
-	                          final QueryProcessor qProc ) {
+	                          final QueryProcessor qProc,
+	                          final ExecutorService execFed,
+	                          final ExecutorService execPlan ) {
```

Comment:
It somehow doesn't feel nice that these two exec.services need to be passed explicitly here. I understand why you do it, but I feel that there should be a nicer solution.

For `execFed`, an better option may be to add a `shutdown()` function to the `FederationAccessManager` interface and call that one in the `shutdown()` function below here.

Similarly for `execPlan`: add a `shutdown()` function to `QueryProcessor` and, in the implementation within `QueryProcessorImpl`, get the exec.service from the `QueryProcContext ctxt` to shut it down.

Reply:
Alright, I'll give that a try!